numba/untyped_passes.py

```diff
@@ -463,3 +468,902 @@ def run_pass(self, state):
+        entry_block.body[idx + 1].value.value = call_get_range_var
+
+        glbls = copy(func_ir.func_id.func.__globals__)
+        inline_closure_call(func_ir, glbls, entry_block, idx, get_range,)
+        kill = entry_block.body.index(assgn)
+        entry_block.body.pop(kill)
+
+        # find the induction variable + references in the loop header
+        # fixed point iter to do this, it's a bit clunky
+        induction_vars = set()
+        header_block = func_ir.blocks[loop.header]
+
+        # find induction var
+        ind = [x for x in header_block.find_exprs('pair_first')]
+        for x in ind:
+            induction_vars.add(func_ir.get_assignee(x, loop.header))
+        # find aliases of the induction var
+        tmp = set()
+        for x in induction_vars:
+            try:  # there's not always an alias, e.g. loop from inlined closure
+                tmp.add(func_ir.get_assignee(x, loop.header))
+            except ValueError:
+                pass
+        induction_vars |= tmp
+        induction_var_names = set([x.name for x in induction_vars])
+
+        # Find the downstream blocks that might reference the induction var
+        succ = set()
+        for lbl in loop.exits:
+            succ |= set([x[0] for x in cfg.successors(lbl)])
+        check_blocks = (loop.body | loop.exits | succ) ^ {loop.header}
+
+        # replace RHS use of induction var with getitem
+        for lbl in check_blocks:
+            for stmt in func_ir.blocks[lbl].body:
+                if isinstance(stmt, ir.Assign):
+                    # check for aliases
+                    try:
+                        lookup = getattr(stmt.value, 'name', None)
+                    except KeyError:
+                        continue
+                    if lookup and lookup in induction_var_names:
+                        stmt.value = ir.Expr.getitem(
+                            iterarg, stmt.value, stmt.loc)
+
+        post_proc = postproc.PostProcessor(func_ir)
+        post_proc.run()
+
+    def run_pass(self, state):
+        func_ir = state.func_ir
+        cfg = compute_cfg_from_blocks(func_ir.blocks)
+        loops = cfg.loops()
+
+        mutated = False
+        for header, loop in loops.items():
+            stat = self.assess_loop(loop, func_ir, state.typemap)
+            if stat:
+                if self._DEBUG:
+                    print("Canonicalising loop", loop)
+                self.transform(loop, func_ir, cfg)
+                mutated = True
+            else:
+                if self._DEBUG:
+                    print("NOT Canonicalising loop", loop)
+
+        func_ir.blocks = simplify_CFG(func_ir.blocks)
+        return mutated
+
+
+@register_pass(mutates_CFG=True, analysis_only=False)
+class LiteralUnroll(FunctionPass):
+    """Implement the literal_unroll semantics"""
+    _name = "literal_unroll"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    def run_pass(self, state):
```

Comment:
Need to decide on whether to put a guard on `literal_unroll` in function `__globals__` here.

Reply:
If it's not too hard to add, it sounds like a good idea

Reply:
done